README.md
# DOS-DRV-ihpfs
HPFS read only driver for DOS

## LICENSE
* GNU GPL V2

## COMPILE TOOLS
* 
 
## AUTHORS
* Veit Kannegieser
* Marcus Better

## LINKS
* http://kannegieser.net/veit/programm/index_e.htm


history.txt
iHPFS Revision History
----------------------

1.28	Fixed version check bug that prevented iHPFS to run under Novell DOS.

1.27	Runs under OpenDOS 7.02.

	Supports up to 32 drives letters (under some versions of OpenDOS).

1.26	Fixed a problem that caused hangs with MSCDEX in DOS 7.

1.25	Fixed file search problem in DOS.

	Changes to partition scanning code.

1.24	Works with Windows 95.

	Several modifications to the partition scanning code. 

	Now supports large IDE drives on more BIOSes through the /B option.

1.23	The partition scanning code has been rewritten from scratch. Most
	notably, this should fix problems with large IDE drives.

	The /B option has been removed, at least for the time being.

	Attempts to read a file starting beyond the end of the file will now
	be handled the way applications expect.

	The driver no longer leaves a small stub in memory after an uninstall,
	except when another TSR loaded after iHPFS prevents this.

	Fixed a bug that would cause a "Divide by zero" error when a partition
	was first accessed under certain conditions.

	Directory searches now imitate the behaviour of DOS more closely.

	Fixed a bug which caused Novell DOS 7 to crash when the cache was
	activated.

1.22	Added support for Novell DOS 7.

	New /B option forces use of extended BIOS functions.

1.21	Added support for large IDE drives (>528 MB).

	Added support for DR DOS 5 and 6.

	Fixed a problem with unusually long paths. If a pathname is too long
	for DOS to handle, the directory tree is cut off at the deepest 
	possible level, and subdirectories at the last level are seen as
	inaccessible files with length zero. 

1.20	iHPFS can now handle several drives in one instance. The command
	line syntax has changed to reflect this. See the documentation
	for details.

	Free space on the iHPFS partitions is now reported.

	Access has been speeded up a bit.

	The . and .. directories no longer exist in the root directory.

	A bug with the /L option in version 1.15 would cause iHPFS to
	report that a file/directory existed when only a partial file name
	was given. This has been corrected.

1.15	Fixed a problem with crashes when running iHPFS with MSD 2.11 and
	some other programs.
	
	Should now work better with fragmented files.

1.14	A new /L option gives access to files with long filenames.

	iHPFS now returns the correct disk size.

	Removed a limitation on the number of simultaneously open files -
	this works better with Windows.

	The . and .. directories are now explicitly returned in directory
	searches, which is necessary for some programs.

1.13	Speeded up the cache a lot.

	The driver may now be uninstalled using the /U option.

	Fixed a bug with directory searches. Now works with TREE.

1.12	Added support for secondary partitions.

1.11	Fixed bugs with directory searches.

1.10	Cache added. The cache uses XMS memory.

	Handles volume labels.

	Won't run under OS/2 DOS.

1.00	First release

killdrv.asm
;
; KillDrv - Drive removal utility
; Copyright (C) 1994-1997 Marcus Better
;
; This program is free software; you can redistribute it and/or modify
; it under the terms of the GNU General Public License as published by
; the Free Software Foundation; either version 2 of the License, or
; (at your option) any later version.
;
; This program is distributed in the hope that it will be useful,
; but WITHOUT ANY WARRANTY; without even the implied warranty of
; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
; GNU General Public License for more details.
;
; You should have received a copy of the GNU General Public License
; along with this program; if not, write to the Free Software
; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
;

	IDEAL
	MODEL	SMALL
	JUMPS
	LOCALS
;========================================================= MACROs
MACRO   Abort   Code
	mov     bx, Code
	call    AbortMsg
ENDM

	STACK
;========================================================= CODE segment
	CODESEG
	STARTUPCODE

PROC	Main
; Write hello message
	mov     ah, 09h
	mov     dx, OFFSET MsgHello
	int     21h
; Check for DR-DOS or Novell DOS
	mov	ax, 4452h
	stc
	int	21h
	jc	@@NotDrDos
; DR-DOS or Novell DOS found. Check version.
	cmp	ah, 10h
	jne	@@errBadDosVer
	cmp	al, 72h		; Novell DOS
	je	@@NovellDOS
	mov	[DrDos], al	; DR-DOS version code
	mov	[CDSSize], 51h	; CDS size different under DR-DOS
	cmp	al, 65h		; DR-DOS 5
	je	@@DosVerOk
	cmp	al, 67h		; DR-DOS 6
	je	@@DosVerOk
	jmp	@@errBadDosVer
@@NovellDOS:
;	mov	[Novell], 1
	jmp	@@DosVerOk
@@NotDrDos:
; Check DOS version
	mov     ax, 3000h
	int     21h
	cmp     al, 4
	jb      @@errBadDOSVer
	cmp     al, 10
	jae	@@errBadDOSVer
@@DosVerOk:

;Parse command line
	call	ParseCmdLine
	cmp	[DriveNo], 0FFh
	jne	@@DrvSpecd
	Abort	1			; Drive not specified
@@DrvSpecd:
; Get list of lists.
	mov	ah, 52h
	int	21h			; ES:BX->List of lists
; Compare drive letter with lastdrive.
	mov	al, [es:bx+21h]	
	cmp	al, [DriveNo]
	ja	@@LastdrvOk
	Abort	2			; Invalid drive letter
@@LastdrvOk:
; Find CDS for drive
	les	bx, [es:bx+16h]		; CDS array
	mov     al, [DriveNo]
	mov     dl, [BYTE CDSSize]
	mul     dl
	add     bx, ax          	; CDS entry for drive
	cmp	[DrDos], 0
	jz	@@DisableMSDOS
	mov	[WORD es:bx+43h], 0
	jmp	@@Disabled
@@DisableMSDOS:
	and	[BYTE PTR es:bx+44h], 03Fh ; Remove drive
@@Disabled:
; Print message and return
	mov	al, [DriveNo]
	add	[RemovedDrv], al
	mov	dx, OFFSET OkMsg
	mov	ah, 9
	int	21h
	mov	ax, 4C00h
	int	21h

@@errBadDosVer:
	Abort   3
ENDP	Main

;---------------------------------------------------------------------
PROC    AbortMsg
	ASSUME  ds:@data,es:NOTHING
	mov     ax, @data
	mov     ds, ax
	shl     bx, 1
	mov     dx, [ErrMsgTbl+bx]
	mov     ah, 9
	int     21h
	mov     ax, bx
	shr     ax, 1
	mov     ah, 4Ch
	int     21h
ENDP    AbortMsg

;---------------------------------------------------------------------
; Parse command line.
PROC    ParseCmdLine STDCALL
	LOCAL   LastByte
	mov	ah, 51h
	int	21h		; Get PSP
	mov	es, bx
	mov     al, [BYTE PTR es:80h]
	cmp     al, 2
	jb      @@Done
	add     al, 80h
	xor     ah, ah
	mov     [LastByte], ax
; Find first non-space character
	mov     di, 81h
@@Next:
	mov     cx, [LastByte]
	sub     cx, di
	jc      @@Done
	inc     cx
	mov     al, ' '
	cmp     al, ' '         ; Set zero flag
	repe scasb
	je      @@Done
	dec     di
; Get char and convert to upper case
	mov     dl, [es:di]
	mov     ax, 6520h
	int     21h
; Check for drive spec 'A'..'Z'
	cmp     dl, 'A'
	jb      @@Parse2
	cmp     dl, 'Z'
	ja      @@Parse2
; Drive spec.
	inc     di
	mov     al, ':'                 ; Match a colon
	scasb
	jne     @@errBadOption
	sub     dl, 'A'                 ; Drive number
	mov     [DriveNo], dl
	jmp     @@Next
@@Parse2:
; Check for switches
	mov     al, '/'
	scasb
	jne     @@errBadOption
; Switch
	cmp     di, [LastByte]
	ja      @@errBadOption
	mov     dl, [es:di]
	inc     di
	mov     ax, 6520h		; Upper case
	int     21h
; Insert any switches here...
	jmp     @@errBadOption

@@errBadOption:
	Abort   0			; Bad cmdline syntax

@@Done:
	ret
ENDP    ParseCmdLine

;========================================================= DATA segment
	DATASEG
DriveNo	DB	0FFh
DrDos	DB	0	; Nonzero if DR-DOS (zero if Novell DOS)
CDSSize	DW	58h	; Size of CDS entry
; Messages
ErrMsgTbl DW	MsgBadOption		; 0 - Bad cmdline syntax
	DW	MsgNoDriveSpec		; 1 - No drive specified
	DW	MsgInvalidDrv		; 2 - Invalid drive letter
	DW	MsgBadDOSVer		; 3 - Bad DOS version

MsgHello DB     "KillDrv            Drive removal utility           "
	DB	"Version 1.01    97-05-03",10,13
	DB      "Copyright (c) 1994-1997 Marcus Better.",10,13,10,13,"$"
MsgBadOption DB "Invalid command line arguments.",10,13,10,13
	DB      "Syntax: KILLDRV d:",10,13
	DB      "where d is the drive letter to be removed.",10,13,"$"
MsgNoDriveSpec DB "No drive letter specified.",10,13,"$"
MsgInvalidDrv DB "Invalid drive letter.",10,13,"$"
MsgBadDOSVer DB	"Wrong DOS version.",10,13,"$"
OkMsg	DB	"Removed drive "
RemovedDrv DB	"A:",10,13,"$"
;========================================================= 
	END

dumbdrv.asm
;
; DumbDrv - Dummy DOS Device Driver
; Copyright (C) 1994-1997 Marcus Better
;
; This program is free software; you can redistribute it and/or modify
; it under the terms of the GNU General Public License as published by
; the Free Software Foundation; either version 2 of the License, or
; (at your option) any later version.
;
; This program is distributed in the hope that it will be useful,
; but WITHOUT ANY WARRANTY; without even the implied warranty of
; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
; GNU General Public License for more details.
;
; You should have received a copy of the GNU General Public License
; along with this program; if not, write to the Free Software
; Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
;

IDEAL
JUMPS
LOCALS

;========================================================= CONSTANTS
; Status codes
stSuccess 	EQU	0100h
stError		EQU	8100h
stUnknownCmd	EQU	8103h
stDrvNotReady	EQU	8102h

MediaID		EQU	0F0h	; Media ID byte - "other"

;========================================================= Driver segment
SEGMENT DrvSeg
	ASSUME cs:DrvSeg,ds:NOTHING,es:NOTHING,ss:NOTHING
; Device header
	DW	0FFFFh, 0	; Link to next driver
	DW	0		; Attributes
	DW	Strategy, Interrupt
	DB	1, 7 DUP (0)	; Number of units

;========================================================= Resident DATA area
; Data area
ReqHdrOfs DW	0		; Offset ot request header
ReqHdrSeg DW	0		; Segment of request header
VolumeID DB	"DUMBDRV", 0
; BIOS Parameter Block
LABEL	BPB
BytesPerSec	DW	512
SecPerClust	DB	2
ResSectors	DW	1
FATs		DB	1
RootDirEnts	DW	32
Sectors		DW	512
Media		DB	0F0h
FATsecs		DW	4
SecPerTrack	DW	9
Heads		DW	2
HiddenSecs	DD	0
HugeSectors	DD	0
pBPB	DW	26 DUP(OFFSET BPB) ; Array of BPB pointers

JmpTbl	DW	Init, MediaChk, BuildBPB, 0
	DW	NotReady, 0, 0, 0
	DW	NotReady, NotReady

;========================================================= STRATEGY routine
PROC	Strategy FAR
	mov	[ReqHdrOfs], bx
	mov	[ReqHdrSeg], es
	ret
ENDP	Strategy

;========================================================= INTERRUPT routine
PROC	Interrupt FAR
	push	ax bx cx dx si di bp ds es
	pushf

	push	cs
	pop	ds
	ASSUME	ds:DrvSeg
	les	bx, [DWORD ReqHdrOfs]
	mov	al, [es:bx+2]	; Function number
	cmp	al, 09h
	ja	@@errUnknownCommand
	xor	ah, ah
	mov	si, ax
	shl	si, 1
	mov	cx, [JmpTbl+si]
	jcxz	@@errUnknownCommand
	call	cx
	jmp	@@Exit

@@errUnknownCommand:
	mov	[WORD es:bx+3], stUnknownCmd
@@Exit:	popf
	pop	es ds bp di si dx cx bx ax
	ret
ENDP	Interrupt

;========================================================= Functions
; Media check
PROC	MediaChk
	mov	[WORD es:bx+3], stSuccess	; Status
	mov	[BYTE es:bx+14], 1		; Media unchanged
	mov	[WORD es:bx+15], OFFSET VolumeID
	mov	[es:bx+17], cs
	ret
ENDP	MediaChk

; Build BPB
PROC	BuildBPB
	mov	[WORD es:bx+3], stSuccess	; Status
	mov	[WORD es:bx+18], OFFSET BPB
	mov	[es:bx+20], cs
	ret
ENDP	BuildBPB

; Return "Drive not ready" error for Read or Write req.
PROC	NotReady
	mov	[WORD es:bx+3], stDrvNotReady
	mov	[WORD es:bx+18], 0		; Sectors read/written
	mov	[WORD es:bx+22], OFFSET VolumeID
	mov	[WORD es:bx+24], cs
	ret	
ENDP	NotReady

LABEL	EndResCode
;========================================================= INIT Code
PROC	Init
; Say hello
	mov	ah, 09h
	mov	dx, OFFSET HelloMsg
	int	21h
; Parse command line
	call	ParseCmdLine
	jc	@@CmdLineError
	or	al, al
	jz	@@CmdLineError			; Zero drives specified
	mov	ah, [es:bx+22]			; First drive number
	add	ah, al
	cmp	ah, 25				; Maximum number of drives
	jna	@@DrivesOk
	mov	al, 26
	sub	al, [es:bx+22]
@@DrivesOk:
	mov	[Drives], al
; Set fields
	mov	[BYTE es:bx+13], al		; Number of units
	mov	[WORD es:bx+3], stSuccess	; Status
	mov	[WORD es:bx+14], OFFSET EndResCode
	mov	[es:bx+16], cs
	mov	[WORD es:bx+18], OFFSET pBPB
	mov	[es:bx+20], cs
	mov	[WORD es:bx+23], 1		; Error message flag
; Print message
	mov	ah, [es:bx+22]			; Drive number
	add	[DrvLetter], ah
	mov	ah, 09h
	mov	dx, OFFSET InstallMsg
	int	21h
	mov	al, [Drives]
	cmp	al, 1
	je	@@Msg1
	add	al, [es:bx+22]
	dec	al
	add	[DrvLetter2], al
	mov	ah, 09h
	mov	dx, OFFSET InstMsg2
	int	21h
	ret
@@Msg1:
	mov	ah, 09h
	mov	dx, OFFSET MsgEnd
	int	21h
	ret
; Bad cmd line syntax - fail init
@@CmdLineError:
	mov	ah, 09h
	mov	dx, OFFSET Syntax
	int	21h
	mov	[BYTE es:bx+13], 0
	mov	[WORD es:bx+14], 0
	mov	[WORD es:bx+13], stError
	ret
ENDP	Init

;---------------------------------------------------------------------
; Parse command line.
PROC    ParseCmdLine
	push	ds
	push	bx cx dx si di
	cld
	lds	si, [es:bx+18]
	ASSUME	ds:NOTHING
; Scan past the filename to the first space
@@EatFilename:
	lodsb
	cmp	al, 0Dh
	je	@@NoParams
	cmp	al, 0Ah
	je	@@NoParams
	cmp	al, ' '
	jne	@@EatFilename
; Eat spaces
@@EatSpaces:
	lodsb
	cmp	al, ' '
	je	@@EatSpaces
	cmp	al, 0Ah
	je	@@NoParams
	cmp	al, 0Dh
	je	@@NoParams
	xor	dx, dx
; Parse one- or two-digit number
	cmp	al, '0'
	jb	@@Error
	cmp	al, '9'
	ja	@@Error
	mov	dl, al
	sub	dl, '0'
	lodsb
	cmp	al, 0Ah
	je	@@Done
	cmp	al, 0Dh
	je	@@Done
	cmp	al, ' '
	je	@@EndParams
	cmp	al, '0'
	jb	@@Error
	cmp	al, '9'
	ja	@@Error
	sub	al, '0'
	mov	dh, dl
	mov	cl, 3
	shl	dh, cl
	shl	dl, 1
	add	dl, dh
	add	dl, al
; Make sure command line ends here
@@EndParams:
	lodsb
	cmp	al, 0Ah
	je	@@Done
	cmp	al, 0Dh
	je	@@Done
	cmp	al, ' '
	je	@@EndParams

@@Error:
	stc
	jmp	@@Exit
@@NoParams:
	mov	ax, 1
	clc
	jmp	@@Exit
@@Done:
	mov	al, dl
	xor	ah, ah
	clc
	jmp	@@Exit

@@Exit:
	pop	di si dx cx bx
	pop	ds
	ret
ENDP    ParseCmdLine

;========================================================= INIT data
HelloMsg DB	10,13,"DumbDrv     Version 1.10     96-01-03", 10, 13
	DB      "Copyright (c) 1994-96, Marcus Better.",10,13,10,13,"$"
InstallMsg DB	"DumbDrv installed as "
DrvLetter DB	"A:","$"
InstMsg2 DB	" - "
DrvLetter2 DB	"A:"
MsgEnd	DB	10,13,"$"

Syntax	DB	"Invalid arguments.",10,13,"$"
Drives	DB	1	; Number of drives to support

ENDS	DrvSeg
	END